File: server/python/app/routes/explanation.py
from flask import Blueprint, request, jsonify
from app.data.sample_explanations import explanation_data
from app.utils.llm_utils import generate_explanation
import logging

explanation_bp = Blueprint("explain", __name__)


@explanation_bp.route("/search", methods=["GET"])
def search():
    """Search for an explanation of a query."""
    query = request.args.get("q", "").lower()
    if not query:
        return jsonify({"error": "No search query provided"}), 400
    model_id = request.args.get("modelId")
    if not model_id:
        return jsonify({"error": "LLM model ID is required"}), 400
    try:
        explanation = generate_explanation(query, model_id)
        return jsonify({"entry": explanation})
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@explanation_bp.route("/autosuggest", methods=["GET"])
def autosuggest():
    """
    Provide autosuggest results for a partial query.

    Parameters:
    - q: The partial query to get suggestions for

    Returns:
    - A list of suggested queries
    """
    prefix = request.args.get("q", "").lower()
    if not prefix:
        return jsonify([])

    # Filter explanation queries that start with the prefix
    suggestions = [
        query for query in explanation_data.keys() if query.startswith(prefix)
    ]
    return jsonify(suggestions[:10])  # Limit to 10 suggestions

File: server/go/routes/explanation.go
package routes

import (
	"log"
	"os"
	"redefine/server/autosuggest"
	"redefine/server/db"
	"redefine/server/llm"
	_ "redefine/server/llm/providers"

	"github.com/gin-gonic/gin"
)

var autosuggestProvider autosuggest.Provider

func init() {

	providerType := os.Getenv("AUTOSUGGEST_PROVIDER")

	var err error
	autosuggestProvider, err = autosuggest.NewProvider(providerType)
	if err != nil {
		log.Printf("Failed to create autosuggest provider: %v. Using default provider.", err)
		autosuggestProvider, _ = autosuggest.NewProvider("")
	} else if providerType != "" {
		log.Printf("Using autosuggest provider: %s", providerType)
	}

	dataSource := os.Getenv("AUTOSUGGEST_DATA_SOURCE")
	if dataSource == "" {
		dataSource = autosuggest.GetDefaultWordsPath()
	}

	log.Printf("Loading words from source: %s", dataSource)
	if err := autosuggestProvider.LoadData(dataSource); err != nil {
		log.Printf("Failed to load words from source: %v", err)
	}
}

func setupExplanationRoutes(api *gin.RouterGroup) {
	explainGroup := api.Group("/explain")

	explainGroup.GET("/search", search)

	explainGroup.GET("/autosuggest", autosuggestHandler)
}

func search(c *gin.Context) {
	query := c.Query("q")
	if query == "" {
		c.JSON(400, gin.H{"error": "No search query provided"})
		return
	}

	modelID := c.Query("modelId")
	if modelID == "" {
		c.JSON(400, gin.H{"error": "LLM model ID is required"})
		return
	}

	promptType := c.Query("promptType")
	if promptType == "" {
		promptType = "general"
	}

	explanation, err := llm.GenerateExplanation(
		query,
		modelID,
		db.GetLLMModelByID,
		func(pType string) (string, error) { return db.GetPromptTemplate(pType) },
		promptType,
	)
	if err != nil {
		log.Printf("Error generating explanation: %v", err)
		c.JSON(500, gin.H{"error": err.Error()})
		return
	}

	c.JSON(200, gin.H{"entry": *explanation})
}

func autosuggestHandler(c *gin.Context) {
	query := c.Query("q")
	limit := 10

	response := autosuggest.HandleRequest(autosuggestProvider, query, limit)

	c.JSON(200, response.Suggestions)
}
